controller/controller_server/utils.py

```diff
@@ -8,42 +10,153 @@
+
+
+def _get_secret_from_secrets(secrets, topologies=None, secret_uid=None):
+    secret_config = secrets.get(config.SECRET_CONFIG_PARAMETER)
+    secret = secrets
+    if secret_config:
+        if secret_uid:
+            secret = get_secret_by_uid(secret_config_string=secret_config, secret_uid=secret_uid)
+        elif topologies:
+            secret, secret_uid = get_secret_by_topologies(secret_config_string=secret_config,
+                                                          node_topologies=topologies)
+        else:
+            raise ValidationException(messages.invalid_secret_message)
+    return secret, secret_uid
+
+
+def _get_array_connection_info_from_secret(secret):
+    user = secret[config.SECRET_USERNAME_PARAMETER]
+    password = secret[config.SECRET_PASSWORD_PARAMETER]
+    array_addresses = secret[config.SECRET_ARRAY_PARAMETER].split(config.PARAMETERS_ARRAY_ADDRESSES_DELIMITER)
     return user, password, array_addresses
 
 
-def get_volume_id(new_volume):
-    return _get_object_id(new_volume)
+def get_array_connection_info_from_secrets(secrets, topologies=None, secret_uid=None):
+    secret, secret_uid = _get_secret_from_secrets(secrets, topologies, secret_uid)
+    user, password, array_addresses = _get_array_connection_info_from_secret(secret)
+    return Secret(user=user, password=password, array_addresses=array_addresses, uid=secret_uid)
+
+
+def get_pool_from_parameters(parameters, secret_uid=None):
+    pools_by_system = parameters.get(config.PARAMETERS_POOLS_BY_SYSTEM)
+    if pools_by_system and secret_uid:
+        return get_pool_by_uid(pools=pools_by_system, uid=secret_uid)
+    return parameters.get(config.PARAMETERS_POOL)
+
+
+def get_volume_id(new_volume, secret_uid):
+    return _get_object_id(new_volume, secret_uid)
 
 
 def get_snapshot_id(new_snapshot):
     return _get_object_id(new_snapshot)
 
 
-def _get_object_id(obj):
+def _get_object_id(obj, secret_uid=None):
+    if secret_uid:
+        return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, secret_uid, obj.id))
     return config.PARAMETERS_OBJECT_ID_DELIMITER.join((obj.array_type, obj.id))
 
 
-def validate_secret(secret):
-    logger.debug("validating secrets")
-    if secret:
-        if not (config.SECRET_USERNAME_PARAMETER in secret and
-                config.SECRET_PASSWORD_PARAMETER in secret and
-                config.SECRET_ARRAY_PARAMETER in secret):
+def _is_imported_string_valid(imported_string):
+    return re.match(config.SECRET_VALIDATION_REGEX, imported_string) and imported_string == imported_string.strip()
+
+
+def _validate_secret_uid(uid):
+    if not _is_imported_string_valid(uid):
+        raise ValidationException(messages.invalid_uid_value_message.format(uid))
+    if len(uid) > config.SECRET_UID_MAX_LENGTH:
+        raise ValidationException(
+            messages.parameter_length_is_too_long.format("secret_uid", uid, config.SECRET_UID_MAX_LENGTH))
+
+
+def _validate_secret(secret):
```

Comment:
I know this is existing code, but we should prefer spec terms over k8s terms, so this dict should be `secrets`, and not `secret`

Reply:
done

Reply:
I still see the parameter name is `secret` (singular)

Reply:
done